src/Network/Transport/TCP.hs

```diff
@@ -1717,7 +1743,9 @@ runScheduledAction (ourEndPoint, theirEndPoint) mvar = do
     handleIOException ex vst = do
       -- Release probing resources if probing.
       forM_ (remoteProbing vst) id
-      tryCloseSocket (remoteSocket vst)
```

Comment:
You're right about this. Must always wait until `handleIncomingMessages` has stopped before posting `EventConnectionLost`.

I've patched it so that only `handleIncomingMessages` will post `EventConnectionLost`, and always just before it closes up and stops receiving. Shutting down the socket here will make that happen, eventually. It's slightly complicated by the special case for closing an end point: the socket will shutdown, and we'll see a premature exit with exception on the socket read, but in that case we must not enqueue `EventConnectionLost`, as there's a network-transport test which checks that they're skipped and `EndPointClosed` comes out instead. That's done by checking the local endpoint state and posting only if it's valid.

Also noticed that I've passed the wrong address to `qdiscEnqueue'`. Should be the peer's address.